### PR TITLE
APP-1844 Create context menu component

### PR DESCRIPTION
**Pull request by ehhong**
Context menu represents the menu itself -- any positioning / floating behavior must be externally managed (see test example)

design https://www.figma.com/file/ERHSYDGTXiEQOVGY8ORg6I/PRIME?type=design&node-id=2703-5151&t=PUd4XMOf36ba4gDT-0

### CODE REVIEW

File: playground/src/context-menu-test.vue [hunk 1]
```diff
@@ -0,0 +1,50 @@
+<script setup lang="ts">
+import { useFloating } from '@floating-ui/vue';
+import { ref } from "vue";
+
+let showMenu = $ref<boolean>(false)
+
+const parent = ref<HTMLElement>();
+const child = ref<HTMLElement>();
+const { floatingStyles } = useFloating(parent, child, {placement: 'bottom-start'});
```
Comment (micheal-parks) [inline playground/src/context-menu-test.vue:9]: I wouldn't recommend testing anything relating to floating UI here. The library itself could break or change, and maintenance would then would be complicated for someone trying to introduce an unrelated change to the library.

Comment (ehhong) [reply playground/src/context-menu-test.vue:9]: what do you think of leaving the floating UI stuff in this test file as an example? I am not writing any playwright tests related to floating UI

File: tests/icon.spec.ts [hunk 2]
```diff
@@ -20,62 +20,62 @@ test('Renders and icon as extra small if size attribute is specified as xs', asy
   page,
 }) => {
   await expect(page.getByTestId('icon-size-xs').locator('i')).toHaveClass(
-    /text-xs/
+    /text-xs/u
```
Comment (ehhong) [inline tests/icon.spec.ts:23]: fixing lint warnings

File: tests/select.spec.ts [hunk 3]
```diff
@@ -23,9 +23,9 @@ test('Given an options attribute, on select the options should be visible and cl
   // within the optionsContainer there should be 3 different options
   expect(await optionsContainer.locator('label').all()).toHaveLength(3);
 
-  await expect(optionsContainer).toHaveText(/one/);
-  await expect(optionsContainer).toHaveText(/two/);
-  await expect(optionsContainer).toHaveText(/three/);
+  await expect(optionsContainer).toHaveText(/one/u);
+  await expect(optionsContainer).toHaveText(/two/u);
+  await expect(optionsContainer).toHaveText(/three/u);
```
Comment (ehhong) [inline tests/select.spec.ts:28]: lint warnings

File: src/elements/context-menu/context-menu.svelte [hunk 4]
```diff
@@ -0,0 +1,11 @@
+<svelte:options immutable tag="v-context-menu" />
+
+<script lang="ts">
+import { addStyles } from '../../lib/index';
+
+addStyles();
+</script>
+
+<div class="py-1 border border-medium shadow-sm max-w-xs">
```
Comment (mcous) [inline src/elements/context-menu/context-menu.svelte:9]: Should this have an `aria-role` of [`menu`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/menu_role)?

File: src/elements/context-menu/context-menu-item.svelte [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+const handleSelect = () => {
+  dispatch('select', { value: label });
+};
+</script>
+
+<div>
+  {#if isDivider}
+    <hr class="border-light mt-1 mb-0.5 drop-shadow-none" />
+  {:else}
```
Comment (mcous) [inline src/elements/context-menu/context-menu-item.svelte:28]: I think `v-context-menu-divider` should be a separate component; it muddies the water a little by overloading the `menu-item`, which is a different role; `separator` vs `menuitem`

File: tests/context-menu.spec.ts [hunk 6]
```diff
@@ -0,0 +1,69 @@
+import { test, expect } from '@playwright/test';
+import { waitForCustomEvent } from './lib/helper.js';
+
+test.beforeEach(async ({ page }) => {
+  await page.goto('/playground/context-menu-test.html');
+});
+
+test('Renders labels, dividers, and icons', async ({ page }) => {
+  const contextMenu = page.getByTestId('context-menu-default');
```
Comment (mcous) [inline tests/context-menu.spec.ts:9]: I think in these tests we should strive to minimize `getByTestId`. I think using it to select to overall component for the given `test` is good, but other usages are questionable

File: tests/context-menu.spec.ts [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+  await expect(contextMenu).toBeVisible();
+  await expect(contextMenu.getByText('label 1')).toBeVisible();
+  await expect(contextMenu.getByTestId('context-menu-divider')).toBeVisible();
```
Comment (mcous) [inline tests/context-menu.spec.ts:12]: An `hr` should have an implicit role of `separator`, so we should use that instead of `test-id`

File: src/elements/context-menu/context-menu-item.svelte [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+export let label: string;
+
+$: isDivider = htmlToBoolean(divider, 'divider');
+
+const handleSelect = () => {
+  dispatch('select', { value: label });
+};
+</script>
+
+<div>
```
Comment (mcous) [inline src/elements/context-menu/context-menu-item.svelte:25]: Should have a `role` of [`menuitem`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/menuitem_role)

File: tests/context-menu.spec.ts [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+import { waitForCustomEvent } from './lib/helper.js';
+
+test.beforeEach(async ({ page }) => {
+  await page.goto('/playground/context-menu-test.html');
+});
+
+test('Renders labels, dividers, and icons', async ({ page }) => {
+  const contextMenu = page.getByTestId('context-menu-default');
+  await expect(contextMenu).toBeVisible();
+  await expect(contextMenu.getByText('label 1')).toBeVisible();
```
Comment (mcous) [inline tests/context-menu.spec.ts:11]: Can these search by `menuitem` role and check the text, instead?

File: tests/context-menu.spec.ts [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+  // renders with icon
+  await expect(contextMenu.getByText('label 3')).toBeVisible();
+  await expect(
+    contextMenu.getByTestId('context-menu-icon').locator('i')
+  ).toBeVisible();
+  await expect(
+    contextMenu.getByTestId('context-menu-icon').locator('i')
+  ).toHaveClass(/icon-trash/u);
+  await expect(
+    contextMenu.locator('i:left-of(:text("label 3"))').first()
```
Comment (mcous) [inline tests/context-menu.spec.ts:24]: Not super familiar with the locator syntax, but could this be `i.icon-trash:left-of...` or similar to collapse these two assertions?

**Review by mcous — COMMENTED** [review]
Styling matches the Figma perfectly, as far as I can tell. Two suggestions for the component itself from an accessibility standpoint:

- Add roles
- Split apart "item" and "divider" into two separate components

With these roles added, you should also be able to reduce usage of `getByTestId` in the tests

File: package.json [hunk 11]
```diff
@@ -89,5 +89,8 @@
     "vite": "4.3.1",
     "vue": "^3.2.45",
     "vue-tsc": "^1.2.0"
+  },
+  "dependencies": {
+    "@floating-ui/vue": "^1.0.0"
```
[the remark this answers is not in the document]
Comment (ehhong) [reply package.json:94]: good catch! thank you

File: tests/context-menu.spec.ts (hunk 10 above)
Comment (ehhong) [reply tests/context-menu.spec.ts:24]: yup, this worked!

**Review by mcous — APPROVED** [review]
Nice!

File: playground/src/context-menu-test.vue [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+      ref="child"
+      :style="floatingStyles"
+      @click="showMenu = false"
+    >
+      <v-context-menu-item 
+        icon="send"
+        label="and this is this child"
+        @select="handleSelect"
+      />
+      <v-context-menu-item divider/>
```
[the remark this answers is not in the document]
Comment (ehhong) [reply playground/src/context-menu-test.vue:60]: oops, thanks for catching!

File: src/elements/context-menu/context-menu-item.svelte [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+export let icon: string;
+export let variant: 'primary' | 'danger' = 'primary';
+export let label: string;
+
+const handleSelect = () => {
+  dispatch('select', { value: label });
+};
+</script>
+
+<div>
```
[the remark this answers is not in the document]
Comment (ehhong) [reply src/elements/context-menu/context-menu-item.svelte:21]: good catch! don't need

File: src/elements/context-menu/context-menu-item.svelte [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+    {/if}
+
+    <p
+      id={label}
+      class={cx('font-normal', {
+        'text-default': variant === 'primary',
+        'text-danger-dark': variant === 'danger',
+      })}
+    >
+      {label}
```
[the remark this answers is not in the document]
Comment (ehhong) [reply src/elements/context-menu/context-menu-item.svelte:45]: ah yea, i like the idea of using IDs and children for content. even when i was leveraging `label` for the `onselect` behavior here, it felt a bit wrong because your display content might be unwieldy and not great for an event id.

i'm going to leave `label` here for now, but agree we should have this conversation and update components as we move to a more atomic design. is there somewhere, maybe in jira, where we can add a reminder?
